cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -2758,6 +2773,438 @@ struct GroupedOneFactory {
+    return Status::OK();
+  }
+
+  Result<Datum> Finalize() override {
+    ARROW_ASSIGN_OR_RAISE(auto values_buffer, values_.Finish());
+    ARROW_ASSIGN_OR_RAISE(auto groups_buffer, groups_.Finish());
+    ARROW_ASSIGN_OR_RAISE(auto null_bitmap_buffer, values_bitmap_.Finish());
+
+    auto groups = UInt32Array(num_args_, groups_buffer);
+    ARROW_ASSIGN_OR_RAISE(
+        auto groupings,
+        Grouper::MakeGroupings(groups, static_cast<uint32_t>(num_groups_), ctx_));
+
+    auto values_array_data = ArrayData::Make(
+        out_type_, num_args_,
+        {has_nulls_ ? std::move(null_bitmap_buffer) : nullptr, std::move(values_buffer)});
+    auto values = MakeArray(values_array_data);
+    return Grouper::ApplyGroupings(*groupings, *values);
+  }
+
+  std::shared_ptr<DataType> out_type() const override { return list(out_type_); }
+
+  ExecContext* ctx_;
+  int64_t num_groups_, num_args_ = 0;
+  bool has_nulls_ = false;
+  TypedBufferBuilder<CType> values_;
+  TypedBufferBuilder<uint32_t> groups_;
+  TypedBufferBuilder<bool> values_bitmap_;
+  std::shared_ptr<DataType> out_type_;
+};
+
+template <typename Type>
+struct GroupedListImpl<Type, enable_if_t<is_base_binary_type<Type>::value ||
+                                         std::is_same<Type, FixedSizeBinaryType>::value>>
+    final : public GroupedAggregator {
+  using Allocator = arrow::stl::allocator<char>;
+  using StringType = std::basic_string<char, std::char_traits<char>, Allocator>;
+  using GetSet = GroupedValueTraits<Type>;
+
+  Status Init(ExecContext* ctx, const std::vector<ValueDescr>&,
+              const FunctionOptions* options) override {
+    ctx_ = ctx;
+    allocator_ = Allocator(ctx_->memory_pool());
+    // out_type_ initialized by GroupedListInit
+    groups_ = TypedBufferBuilder<uint32_t>(ctx_->memory_pool());
+    values_bitmap_ = TypedBufferBuilder<bool>(ctx_->memory_pool());
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
+    return Status::OK();
+  }
+
+  Status Consume(const ExecBatch& batch) override {
+    const auto values_array_data = batch[0].array();
+    int64_t num_values = values_array_data->length;
+    int64_t offset = values_array_data->offset;
+
+    const auto groups_array_data = batch[1].array();
+    DCHECK_EQ(groups_array_data->offset, 0);
+    const auto* groups = groups_array_data->GetValues<uint32_t>(1, 0);
+    RETURN_NOT_OK(groups_.Append(groups, num_values));
+
+    if (batch[0].null_count() == 0) {
+      RETURN_NOT_OK(values_bitmap_.Append(num_values, true));
+    } else {
+      const uint8_t* values_bitmap = values_array_data->buffers[0]->data();
+      RETURN_NOT_OK(GroupedValueTraits<BooleanType>::AppendBuffers(
+          &values_bitmap_, values_bitmap, offset, num_values));
+    }
+    return AppendBuffers<Type>(batch);
+  }
+
+  template <typename T = Type>
+  enable_if_t<is_base_binary_type<T>::value, Status> AppendBuffers(
```

Comment:
Is templating like this okay?

Reply:
So, the key insight is that FixedSizeBinaryType should actually go with all the other fixed-size types, not the string type. It only has a validity buffer and a data buffer, just like other fixed-size types. The only snag is that instead of using templates/casting to figure out how much data to copy, you have to use the type instance to know. That's what copy_data_utils.h also does, and why I suggest passing the type to AppendBuffers.

…I guess that gets hard with VisitGroupedValues, though, I think you can still make it work (you would just not use VisitGroupedValues, you'd visit the group ID array and work with offsets and uint8_t* only for fixed size types). However, I think we're getting into the weeds. Let's return it to how it was before, and fix up the cases of AppendBuffers that weren't working. And then we can think about if we want to attempt further generalizations.

Reply:
FixedSizeBinaryType doesn't have an offsets buffer. That's why it fits with the fixed-size types. It _is_ for all intents and purposes the same as an array of integers - it's just that the element width isn't fixed at compile time, and you don't need VisitGroupedValues or anything for the fixed size types.

Reply:
Right now we use `TypedBufferBuilder<CType>` for fixed-width ctypes right now. How would we extend this to `FixedSizeBinaryType`? Maybe using the basic `BufferBuilder` and doing the reserving/appending similar to `CopyData`? Should I go ahead and try and integrate `FixedSizeBinaryType` with other fixed-width types?

Reply:
Everything would work on a `TypedBufferBuilder<uint8_t>`, and would work like CopyData, yes. But I think let's just clean this up since it works already. We can file a separate Jira if you'd like to keep looking at optimizations (I think this would be faster for Decimal/FixedSizeBinary but we should have benchmarks to be sure)